app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsViewModel.kt

```diff
@@ -0,0 +1,505 @@
+            FallibleUiAction.ClearNotifications -> ClearNotifications(exception)
+        }
+    }
+}
+
+@OptIn(ExperimentalCoroutinesApi::class, FlowPreview::class)
+class NotificationsViewModel @Inject constructor(
+    private val repository: NotificationsRepository,
+    private val preferences: SharedPreferences,
+    private val accountManager: AccountManager,
+    private val timelineCases: TimelineCases,
+    private val eventHub: EventHub
+) : ViewModel() {
+
+    val uiState: StateFlow<UiState>
+
+    /** Flow of changes to statusDisplayOptions, for use by the UI */
+    val statusDisplayOptions: StateFlow<StatusDisplayOptions>
+
+    val pagingData: Flow<PagingData<NotificationViewData>>
+
+    /** Flow of user actions received from the UI */
+    private val uiAction = MutableSharedFlow<UiAction>()
+
+    /** Flow of successful action results */
+    // Note: These are a SharedFlow instead of a StateFlow because success or error state does not
+    // need to be retained. A message is shown once to a user and then dismissed. Re-collecting the
+    // flow (e.g., after a device orientation change) should not re-show the most recent success or
+    // error message, as it will be confusing to the user.
+    val uiSuccess = MutableSharedFlow<UiSuccess>()
+
+    /** Flow of transient errors for the UI to present */
+    val uiError = MutableSharedFlow<UiError>()
+
+    /** Accept UI actions in to actionStateFlow */
+    val accept: (UiAction) -> Unit = { action ->
+        viewModelScope.launch { uiAction.emit(action) }
+    }
+
+    init {
+        // Handle changes to notification filters
+        val notificationFilter = uiAction
+            .filterIsInstance<InfallibleUiAction.ApplyFilter>()
+            .distinctUntilChanged()
+            // Save each change back to the active account
+            .onEach { action ->
+                Log.d(TAG, "notificationFilter: $action")
+                accountManager.activeAccount?.let { account ->
+                    account.notificationsFilter = serialize(action.filter)
+                    accountManager.saveAccount(account)
+                }
+            }
+            // Load the initial filter from the active account
+            .onStart {
+                emit(
+                    InfallibleUiAction.ApplyFilter(
+                        filter = deserialize(accountManager.activeAccount?.notificationsFilter)
+                    )
+                )
+            }
+
+        // Save the visible notification ID
+        viewModelScope.launch {
+            uiAction
+                .filterIsInstance<InfallibleUiAction.SaveVisibleId>()
+                .distinctUntilChanged()
+                .collectLatest { action ->
+                    Log.d(TAG, "Saving visible ID: ${action.visibleId}")
+                    accountManager.activeAccount?.let { account ->
+                        account.lastNotificationId = action.visibleId
```

Comment:
Isn't this the same attribute used to limit which (push) notifications are shown in the system ui? Using it like this will cause already read notifications to reappear.

Reply:
It won't, but there's a different bit of weirdness.

NotificationFetcher (untouched in this PR) uses the Mastodon marker API to store the notification high water mark.

If it detects that `lastNotificationId` is less than the marker it sets `lastNotificationId` to the marker value. Resetting the user's position back to the new notifications (`NotificationFetcher.fetchNotifications()`).

I've got an idea for how to fix reading positions generally, in https://github.com/tuskyapp/Tusky/pull/3271 (pretty view: https://github.com/tuskyapp/Tusky/blob/2b24e3b14e639e9a0a8591624893a030d80267b8/doc/SaveReadingPosition.md), but it's outside the scope of this PR. Feedback on it is welcome.